Async iteration and generators (#236)
javascript-tutorial/uk.javascript.info: 1-js/12-generators-iterators/2-async-iterators-generators/article.md

# Асинхронні ітератори та генератори

Асинхронні ітератори дозволяють нам перебирати дані, які надходять асинхронно, на вимогу. Наприклад, коли ми завантажуємо щось частинами через мережу. А асинхронні генератори роблять це ще зручніше.

Давайте спочатку розглянемо простий приклад, щоб зрозуміти синтаксис, а потім розглянемо реальний варіант використання.

## Пригадування ітераторів

Згадаймо тему про ітератори.

> Ідея полягає в тому, що ми маємо об’єкт, наприклад, `range` тут:

Reviewer: ```suggestion
Ідея полягає в тому, що ми маємо об’єкт, наприклад, `range`:
```

```js
let range = {
  from: 1,
  to: 5
};
```

...І ми хотіли б використовувати для нього цикл `for..of`, наприклад, `for(value of range)`, щоб отримати значення від `1` до `5`.

> Іншими словами, ми хочемо додати до об’єкта *можливість ітерації*.

Reviewer: ```suggestion
Інакше кажучи, ми хочемо додати до об’єкта *можливість ітерації*.
```
@Nordtonito http://slovopedia.org.ua/30/53382/27135.html

Це можна реалізувати за допомогою спеціального методу з назвою `Symbol.iterator`:

- Цей метод викликається конструкцією `for..of`, коли запускається цикл, і він повинен повернути об’єкт із методом `next`.

> - На кожній ітерації метод `next()` викликається для наступного значення.

Reviewer: ```suggestion
- На кожній ітерації викликається метод `next()`, щоб отримати наступне значення.
```

- `next()` має повертати значення у формі `{done: true/false, value:<loop value>}`, де `done:true` означає кінець циклу.

> Ось реалізація ітеративного `range`:

Reviewer: ```suggestion
Ось реалізація об’єкта `range`, що перебирається:
```
@Nordtonito https://github.com/javascript-tutorial/uk.javascript.info/issues/8#issuecomment-939801854

};

for(let value of range) {

>   alert(value); // 1 потім 2, потім 3, потім 4, потім 5

Reviewer: ```suggestion
  alert(value); // 1, потім 2, потім 3, потім 4, потім 5
```

}
```

Якщо щось незрозуміло, відвідайте розділ [](info:iterable), у ньому наведено всю інформацію про звичайні ітератори.

## Асинхронні ітератори

Асинхронна ітерація потрібна тоді, коли значення надходять асинхронно: після `setTimeout` або іншого виду затримки.

Найпоширеніший випадок полягає в тому, що об’єкту потрібно зробити мережевий запит для доставлення наступного значення. Реальний приклад цього ми побачимо трохи пізніше.

> Щоб зробити об’єкт повторюваним асинхронно:

Reviewer: ```suggestion
Для того, щоб перебирати об’єкт асинхронно:
```

1. Використовуйте `Symbol.asyncIterator` замість `Symbol.iterator`.

> 2. Метод `next()` повинен повертати проміс (що має виконуватися з наступним значенням).

Reviewer: ```suggestion
2. Метод `next()` повинен повертати проміс (який має містити наступне значення).
```

>    - Ключове слово `async` обробляє це, ми можемо просто створити `async next()`.

Reviewer: ```suggestion
   - Це робиться за допомогою ключового слова `async`. Ось так: `async next()`.
```

3. Щоб перебрати такий об’єкт, ми повинні використовувати цикл `for await (let item of iterable)`.
   - Зверніть увагу на слово `await`.

> Як початковий приклад, створімо ітеративний об’єкт `range`, подібний до попереднього, але тепер він повертатиме значення асинхронно, по одному в секунду.

Reviewer: ```suggestion
Як початковий приклад, створімо об’єкт `range`, що перебирається. Він буде подібний до попереднього, проте повертатиме значення асинхронно -- по одному за секунду.
```

> Все, що нам потрібно зробити, це виконати кілька замін у коді вище:

Reviewer: ```suggestion
Все, що нам потрібно зробити, це зробити кілька замін у коді вище:
```

})()
```

Як бачимо, структура схожа на звичайні ітератори:

> 1. Щоб зробити об’єкт асинхронно ітераційним, він повинен мати метод `Symbol.asyncIterator` `(1)`.

Reviewer: ```suggestion
1. Щоб об’єкт перебирався асинхронно, він повинен мати метод `Symbol.asyncIterator` `(1)`.
```

Ось невелика таблиця з відмінностями:

|       | Ітератори | Асинхронні ітератори |
|-------|-----------|-----------------|
| Метод для забезпечення ітератора | `Symbol.iterator` | `Symbol.asyncIterator` |
| `next()` повертає              | будь-яке значення         | `Promise`  |
| для циклу використовуйте                          | `for..of`         | `for await..of` |

````warn header="Синтаксис оператора розширення `...` не працює асинхронно"
Функції, що вимагають звичайні синхронні ітератори, не працюють з асинхронними.

Наприклад, не буде працювати синтаксис розширення:
```js
alert( [...range] ); // Помилка, немає Symbol.iterator
```

Це природно, оскільки він очікує знайти `Symbol.iterator`, а не `Symbol.asyncIterator`.

Це також стосується `for..of`: синтаксис без `await` потребує `Symbol.iterator`.
````

## Пригадування генераторів

Тепер згадаймо генератори, оскільки вони дозволяють зробити ітеративний код значно коротшим. У більшості випадків, коли хочемо створити ітератор, ми використовуємо генератори.

> Для простоти, опускаючи деякі важливі речі, це "функції, які генерують (yield) значення". Вони детально описані в главі [](info:generators).

Reviewer: ```suggestion
Для простоти, опускаючи деякі важливі речі, це "функції, які генерують (yield) значення". Вони детально описані в розділі [](info:generators).
```

> Генератори позначаються ярликом `function*` (зверніть увагу на зірочку) і використовують `yield`, щоб генерувати значення, тоді ми можемо використовувати `for..of`, щоб перейти до них.

Reviewer: ```suggestion
Генератори позначаються міткою `function*` (зверніть увагу на зірочку) і використовують `yield`, щоб генерувати значення, тоді ми можемо використовувати `for..of`, щоб перейти до них.
```

}

for(let value of generateSequence(1, 5)) {
  alert(value); // 1, потім 2, потім 3, потім 4, потім 5
}
```

> Як ми вже знаємо, щоб зробити об’єкт ітеративним, ми повинні додати до нього `Symbol.iterator`.

Reviewer: ```suggestion
Як ми вже знаємо, щоб об’єкт міг перебиратися, ми повинні додати до нього `Symbol.iterator`.
```

```js
let range = {
  from: 1,
  to: 5,
*!*
  [Symbol.iterator]() {
    return <об’єкт з next, щоб зробити діапазон ітерабельним>
  }
*/!*
}
```

> Звичайною практикою для `Symbol.iterator` є повернення генератора, це робить код коротшим, як ви можете бачити:

Reviewer: ```suggestion
Звичайною практикою для `Symbol.iterator` є повернення генератора. Як ви можете бачити, це робить код коротшим:
```

```js run
let range = {
  from: 1,
  to: 5,

  *[Symbol.iterator]() { // скорочення для [Symbol.iterator]: function*()
    for(let value = this.from; value <= this.to; value++) {
      yield value;
    }
  }
};

for(let value of range) {
  alert(value); // 1, потім 2, потім 3, потім 4, потім 5
}
```

Будь ласка, перегляньте розділ [](info:generators), якщо вам потрібно більше деталей.

У звичайних генераторах ми не можемо використовувати `await`. Усі значення мають надходити синхронно, як того вимагає конструкція `for..of`.

Що, якщо ми хочемо генерувати значення асинхронно? З мережевих запитів, наприклад.

Перейдімо до асинхронних генераторів, щоб зробити це можливим.

> ## Асинхронні генератори (заключно)

Reviewer: ```suggestion
## Асинхронні генератори (нарешті)
```

  let generator = generateSequence(1, 5);
  for *!*await*/!* (let value of generator) {
    alert(value); // 1, потім 2, потім 3, потім 4, потім 5 (із затримкою між ними)
  }

})();
```

Оскільки генератор є асинхронним, ми можемо використовувати всередині нього `await`, проміси, виконувати мережеві запити тощо.

````smart header="Різниця під капотом"

> Технічно, якщо ви досвідчений читач, який пам’ятає деталі про генератори, в них є і внутрішня різниця.

Reviewer: ```suggestion
Якщо ви досвідчений читач і пам’ятаєте деталі про генератори, то знаєте, що технічно в них є внутрішня різниця.
```

Для асинхронних генераторів метод `generator.next()` є асинхронним, він повертає проміси.

У звичайному генераторі ми використовуємо `result = generator.next()`, щоб отримати значення. В асинхронному генераторі ми повинні додати `await`, наприклад:

```js
result = await generator.next(); // result = {value: ..., done: true/false}
```
Ось чому асинхронні генератори працюють з `for await...of`.
````

### Асинхронно ітераційний range

Звичайні генератори можна використовувати як `Symbol.iterator`, щоб зробити код ітерації коротшим.

Подібно до цього, асинхронні генератори можна використовувати як `Symbol.asyncIterator` для реалізації асинхронної ітерації.

> Наприклад, ми можемо змусити об’єкт `range` генерувати значення асинхронно, раз на секунду, замінивши синхронний `Symbol.iterator` на асинхронний `Symbol.asyncIterator`:

Reviewer: ```suggestion
Наприклад, ми можемо змусити об’єкт `range` генерувати значення асинхронно, раз за секунду, замінивши синхронний `Symbol.iterator` на асинхронний `Symbol.asyncIterator`:
```

```js run
let range = {
  from: 1,
  to: 5,

  // цей рядок такий самий, як [Symbol.asyncIterator]: async function*() {
*!*
  async *[Symbol.asyncIterator]() {
*/!*
    for(let value = this.from; value <= this.to; value++) {

>       // робимо паузу між значеннями, чекаємо будь-що

Reviewer: ```suggestion
      // робимо паузу між значеннями, чекаємо на щось
```

(async () => {

  for *!*await*/!* (let value of range) {
    alert(value); // 1, потім 2, потім 3, потім 4, потім 5
  }

})();
```

Тепер значення надходять із затримкою в 1 секунду між ними.

```smart

> Технічно ми можемо додати до об’єкта як `Symbol.iterator`, так і `Symbol.asyncIterator`, тому він ітерується як синхронно (`for..of`), так і асинхронно (`for await..of`).

Reviewer: ```suggestion
Технічно ми можемо додати до об’єкта як `Symbol.iterator`, так і `Symbol.asyncIterator`, тому він може перебиратися як синхронно (`for..of`), так і асинхронно (`for await..of`).
```

Але на практиці це було б дивно.
```

## Приклад із реальної практики: посторінкові дані

> Поки що ми бачили основні приклади, щоб зрозуміти. Тепер розгляньмо реальний випадок використання.

Reviewer: ```suggestion
Поки що ми бачили базові приклади, щоб зрозуміти. Тепер розгляньмо реальний випадок використання.
```

Існує багато онлайн-сервісів, які надають дані посторінково. Наприклад, коли нам потрібен список користувачів, запит повертає попередньо визначену кількість (наприклад, 100 користувачів) — "одну сторінку" та надає URL-адресу наступної сторінки.

Цей підхід дуже поширений. Мова йде не про користувачів, а про що завгодно.

Наприклад, GitHub дозволяє нам отримувати коміти таким же чином, розбитими на сторінки:

- Ми повинні зробити запит за допомогою `fetch` у вигляді `https://api.github.com/repos/<repo>/commits`.
- У відповіді прийде JSON з 30 комітами, а також з посиланням на наступну сторінку в заголовку `Link`.

> - Тоді ми зможемо використати це посилання для наступного запиту, щоб отримати більше комітів чи т.п.

Reviewer: ```suggestion
- Тоді ми зможемо використати це посилання для наступного запиту, щоб отримати більше комітів і так далі.
```

Для нашого коду ми хотіли б мати простіший спосіб отримати коміти.

> Створімо функцію `fetchCommits(repo)`, яка отримує коміти для нас, надаючи запити, коли це необхідно. І нехай вона піклується про всі речі щодо розбиття на сторінки. Для нас це буде проста асинхронна ітерація `for await..of`.

Reviewer: ```suggestion
Створімо функцію `fetchCommits(repo)`, яка отримує коміти для нас, надсилаючи запити, коли це необхідно. І нехай вона піклується про всі речі щодо розбиття на сторінки. Для нас це буде проста асинхронна ітерація `for await..of`.
```

Author: Дякую за корекцію перекладу!
